lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -0,0 +1,1427 @@
+      auto index = bodyBuilder.create<LLVM::ConstantOp>(
+          op->getLoc(), i32Ty, rewriter.getI32IntegerAttr(i));
+      auto bitcast = bodyBuilder.create<LLVM::GEPOp>(
+          op->getLoc(), i32Ty.getPointerTo(), procOp.getArgument(2),
+          ArrayRef<Value>({index}));
+      auto load = bodyBuilder.create<LLVM::LoadOp>(op->getLoc(), bitcast);
+
+      // remap i-th original argument to the loaded value
+      final.remapInput(i + 3, load.getResult());
+    }
+
+    rewriter.applySignatureConversion(&procOp.getBody(), final);
+
+    // converted entity signature
+    auto funcTy = LLVM::LLVMType::getFunctionTy(
+        voidTy, {i8PtrTy, stateTy.getPointerTo(), i32Ty.getPointerTo()}, false);
+    // create the llvm function
+    auto llvmFunc = rewriter.create<LLVM::LLVMFuncOp>(op->getLoc(),
+                                                      procOp.getName(), funcTy);
+
+    // inline the entity region in the new llvm function
+    rewriter.inlineRegionBefore(procOp.getBody(), llvmFunc.getBody(),
+                                llvmFunc.end());
+
+    insertPersistence(typeConverter, rewriter, &getDialect(), op->getLoc(),
+                      procOp, stateTy, llvmFunc);
+
+    rewriter.eraseOp(op);
+
+    return success();
+  }
+};
+
+/// Convert an `llhd.halt` operation to llvm dialect. This zeroes out all the
+/// senses and returns, effectively making the process unable to be invoked
+/// again.
+struct HaltOpConversion : public ConvertToLLVMPattern {
+  explicit HaltOpConversion(MLIRContext *ctx, LLVMTypeConverter &typeConverter)
+      : ConvertToLLVMPattern(HaltOp::getOperationName(), ctx, typeConverter) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto i1Ty = LLVM::LLVMType::getInt1Ty(&getDialect());
+    auto i32Ty = LLVM::LLVMType::getInt32Ty(&getDialect());
+
+    auto llvmFunc = op->getParentOfType<LLVM::LLVMFuncOp>();
+    auto procState = llvmFunc.getArgument(1);
+    auto senseTableTy = procState.getType()
+                            .cast<LLVM::LLVMType>()
+                            .getPointerElementTy()
+                            .getStructElementType(2)
+                            .getPointerElementTy();
+
+    // get senses ptr from process state argument
+    auto zeroC = rewriter.create<LLVM::ConstantOp>(
+        op->getLoc(), i32Ty, rewriter.getI32IntegerAttr(0));
+    auto twoC = rewriter.create<LLVM::ConstantOp>(
+        op->getLoc(), i32Ty, rewriter.getI32IntegerAttr(2));
+    auto sensePtrGep = rewriter.create<LLVM::GEPOp>(
+        op->getLoc(), senseTableTy.getPointerTo().getPointerTo(), procState,
+        ArrayRef<Value>({zeroC, twoC}));
+    auto sensePtr = rewriter.create<LLVM::LoadOp>(
+        op->getLoc(), senseTableTy.getPointerTo(), sensePtrGep);
+
+    // zero out all the senses flags
+    for (unsigned int i = 0; i < senseTableTy.getArrayNumElements(); i++) {
```

Comment:
Some nits: "unsigned int" -> "size_t", and no need to evaluate `senseTableTy.getArrayNumElements()` each time through the loop, also generally prefer `++i` to `i++`.  Prefer the pattern:

```
for (size_t i = 0, e = senseTableTy.getArrayNumElements(); i != e; ++i) {
...
```

This occurs a few places in this patch.